### PR TITLE
Generalized profiling scripts

### PR DESCRIPTION
**Pull request by bstefanuk**
**Summary:**

Profiling Tensile's applications, e.g. TensileCreateLibrary, is crucial to understand the runtime statistics required to make meaningful build-time improvements. However, depending on the parameters provided these can take hours to run, making it desirable to run nightly profiles under various contraints, including OS, branch, number of jobs, and target architecture.

**Outcomes:**

The scripts added in this PR systematize job scheduling to enable nightly scripts for the general user.

**Testing and Environment:**

Functional testing only via script invocations.

**Usage documentation**

```
$ ./scripts/init-cron-nightly.sh --help
Set up cron table for nightly Tensile profiling reports

Usage: ./scripts/init-cron-nightly.sh [--build-id=<id>] [--tensile-path=<path>]

Parameters:
  --build-id: The target docker build ID [default: 14543]
  --tensile-path: Path to root directory of Tensile [default: /home/bstefanu/dev/Tensile]

Example:
  ./scripts/init-cron-nightly.sh --build-id=14354 --tensile-path='path/to/tensile'
```

```
$ ./scripts/profile-tcl.sh --help
Run grid-based profiling analysis for TensileCreateLibrary under variable inputs

Usage: ./scripts/profile-tcl.sh --build-id=<build-id> [--branch=<branch>] [--arch=<arch>] [--compiler=<compiler>]

Parameters:
  --build-id: The target docker build ID
  --branch: The target branch [default: develop]
  --arch: Target Gfx architecture(s) [default: gfx900]
  --compiler: HIP-enabled compiler (must be in PATH) [default: amdclang++]

Example:
  ./scripts/profile-tcl.sh --build-id=12345 --branch=develop --arch='gfx90a'

Dependencies:
  docker: Docker is implicitly called and may install images
```

```
$ ./scripts/run-tcl.sh --help
Run TensileCreateLibrary with timestamped log and build directory

Usage: ./scripts/run-tcl.sh --tensile-path=<tensile-path> --logic-path=<logic-path> --jobs=<jobs> [--arch=<arch>] [--compiler=<compiler>]

Parameters:
  --tensile-path: Path to root directory of Tensile
  --logic-path: Path to directory containing logic files
  --jobs: Number of concurrent processes to use
  --arch: Target Gfx architecture(s) [default: gfx900]
  --compiler: HIP-enabled compiler (must be in PATH) [default: amdclang++]

Example:
  ./scripts/run-tcl.sh --tensile-path=/mnt/host/Tensile --logic-path=/mnt/host/Logic --jobs=16
```

### CODE REVIEW

File: scripts/init-cron-nightly.sh [hunk 1]
```diff
@@ -0,0 +1,14 @@
+H1='\033[0;31m'
+H2='\033[0;32m'
+NC='\033[0m' # No color
+
+if crontab -l | grep -q 'run-profile'; then
+    echo -e "$H1> Error: cron job with same command already exists.$NC"
+    echo -e "$H1> Clean your crontab manually with 'crontab -e' and rerun.$NC"
+    echo -e "$H1> Conflict line:\n>   `crontab -l | grep run-profile`$NC"
+    exit 1
+else
```
Comment (bstefanuk) [inline scripts/init-cron-nightly.sh:10]: Add a usage section for this script.

File: scripts/run-profile.sh [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+    echo ""
+    echo "Parameters:"
+    echo "  --build-id: The target docker build ID"
+    echo "  --branch: The target branch"
+    echo "  --archs: Target Gfx architecture(s)"
+    echo "  --compiler: HIP-enabled compiler (must be in PATH)"
+    echo ""
+    echo "Example:"
+    echo "  $0 --build-id=12345 --branch=develop --archs='gfx90a'"
+}
```
Comment (bstefanuk) [inline scripts/run-profile.sh:16]: Specify defaults for optional parameters.

File: scripts/run-tcl.sh [hunk 3]
```diff
@@ -0,0 +1,53 @@
+#!/bin/bash
+
+usage() {
+    echo "HELP COMMAND FOR RUN TCL"
+}
```
Comment (bstefanuk) [inline scripts/run-tcl.sh:5]: Update this usage function with appropriate details.

File: scripts/init-cron-nightly.sh [hunk 4]
```diff
@@ -0,0 +1,58 @@
+H1='\033[0;31m'
+H2='\033[0;32m'
+NC='\033[0m' # No color
+
+# Variables
+build_id="14543"
+tensile_path="`pwd`"
+
+# Constants
+script_to_run="profile-tcl.sh"
```
Comment (ellosel) [inline scripts/init-cron-nightly.sh:10]: How would we reuse this or run multiple scripts?

File: scripts/profile-tcl.sh [hunk 5]
```diff
@@ -0,0 +1,98 @@
+#!/bin/bash
+
+# Variables
+build_id=""
+branch="develop"
+arch="gfx900"
+compiler="amdclang++"
+
+# Constants
+base_image="compute-artifactory.amd.com:5000/rocm-plus-docker/compute-rocm-dkms-no-npi-hipclang"
```
Comment (ellosel) [inline scripts/profile-tcl.sh:10]: Consider removing and require setting it in the users environment.

File: scripts/profile-tcl.sh [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+}
+
+find_tensile() {
+  local query="*/Tensile/setup.py"
+  local os_tag=$1
+  # If several Tensile projects are found, use the first one
+  docker run \
+    --rm \
+    --volume="$HOME:/mnt/host" \
+    "$base_image:$build_id$os_tag" bash -c "find /mnt/host -path $query -exec dirname {} \; 2>&1 | head -n1"
```
Comment (ellosel) [inline scripts/profile-tcl.sh:43]: I can't figure out if users can specify the os_tag or not

File: scripts/profile-tcl.sh (hunk 5 above)
Comment (bstefanuk) [reply scripts/profile-tcl.sh:10]: Sounds good, I'll setup an env var called something like REGISTRY_IMAGE

File: scripts/init-cron-nightly.sh (hunk 4 above)
Comment (bstefanuk) [reply scripts/init-cron-nightly.sh:9]: To re-use this with other script, we can extend this to accept a script as aa command line parameter. For now, this is the only script we intend to use so leaving it hard coded seems like an acceptable solution.

**Review by ellosel — APPROVED** [review]

**Pull request merged**